File: lab10/md5.py
import hashlib

msg	= b"Bryn Mawr"
digest	= hashlib.new("md5")#	creates	an	instance	of	MD5
digest.update(msg)#adds	msg	to	hash	value	computation
digest.digest()#	completes	computation,	returns	result
#print(digest.digest()) -->b'\x03\xb8qF\xec\x1c]\xd0<x\x986\xdaM\xb7\xc2'
#print(digest.hexdigest())-->03b87146ec1c5dd03c789836da4db7c2
#print(hashlib.algorithms_available)
def	hashAFile(fileName,	algorithm):
	#Hash	the	contents	of	the	file,	filename	using	algorithm,
    #and	return	its	hash	value	as	a	hex	string.
    bufferSize	=	1024
    digest	=	hashlib.new(algorithm)
    with	open(fileName,	"rb")	as	inStream:
        buffer = inStream.read(bufferSize)#read bufferSize bytes
        while len(buffer):
            digest.update(buffer)	 #	add	to	digest
            buffer	=	inStream.read(bufferSize)
    return digest.hexdigest()


def hashString(message, algorithm):
    msg2 = message
    digest2 = hashlib.new(algorithm)
    digest2.update(msg2)
    #print(digest2.digest())
    return digest2.hexdigest()

msg	=	b"It was the best of times, it was the worst of times."
hashValue	=	hashString(msg,	"MD5")
print(hashValue," ",msg)
hashValue2	=	hashString(msg,	"SHA256")
print(hashValue2," ",msg)
name1 = "TwoCities.txt"
name2 = "Anassa.txt"
name3 = "moby10b.txt"
name4 = "SpongeBob.png"
fileValue = hashAFile(name1,"MD5")
print(fileValue," ",name1)
fileValue2 = hashAFile(name2,"MD5")
print(fileValue2," ",name2)
fileValue3 = hashAFile(name3,"MD5")
print(fileValue3," ",name3)
fileValue4 = hashAFile(name4,"MD5")
print(fileValue4," ",name4)

fileValueA = hashAFile(name1,"SHA256")
print(fileValueA," ",name1)
fileValueA2 = hashAFile(name2,"SHA256")
print(fileValueA2," ",name2)
fileValueA3 = hashAFile(name3,"SHA256")
print(fileValueA3," ",name3)
fileValueA4 = hashAFile(name4,"SHA256")
print(fileValueA4," ",name4)

File: lab10/md5.java
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.io.UnsupportedEncodingException;
import java.io.*;
public class md5{
    static String	hashString(String	message,	String	algorithm)	{
			//	Compute	the	hash	value	of	message	using	algorithm	and
			//	return	a	string	representation	of	it.
			byte[]	hashedBytes	= null;	 //	will	store	the	hash	value	of	message
			try	{
						//	instantiate	the	specified	algorithm.	
						//	It	may	not	exist,	thus	the	try-catch
						MessageDigest	digest = MessageDigest.getInstance(algorithm);
						
						//	Compute	the	hash	value	of	message
						hashedBytes	=	digest.digest(message.getBytes("UTF-8"));
            }catch(NoSuchAlgorithmException|UnsupportedEncodingException e)	{
						e.printStackTrace();
			}
			//	Convert	hash	value	(in	byte[])	to	a	hex	String	and	return	result
			return	bToH(hashedBytes);
    }	//	hashString()
    static String	bToH(byte[]	value)	{
			//	Converts	value	to	a	string	of	hex	digits
			StringBuilder	sb	=	new	StringBuilder(value.length*2);	
			//	Why,length*2	?	See	Question	3.
			for	(byte	b	:	value)
						sb.append(String.format("%02x",	b));
			return	sb.toString();
    }	//	bToH()
    String	hashAFile(String filename,	String	algorithm)	{
			//	Hash	the	contents	of	the	file,	
			//	fileName	and	return	its	hash	value	as	a	hex	string.
			byte[]	hashedBytes	=null;					//	the	result
			try	{
				//	Open	the	file
				FileInputStream	inStream	=	new	FileInputStream(filename);
                //	Instantiate	a	digest	with	the	algorithm
						MessageDigest	digest	=	MessageDigest.getInstance(algorithm);
						//	Define	input	file	chunk	buffer	(of	1024	bytes)
						byte[]	buffer	=	new	byte[1024];
						int	bytesRead	=	-1;	//	counts	how	many	bytes	were	read
						while	((bytesRead	= inStream.read(buffer))!= -1)	{
									//	there	are	bytes	in	input	file	to	process
									//	supply	the	chunk	to	digest
									digest.update(buffer,0,bytesRead);
						}
						//	finalize	computation
						hashedBytes	=	digest.digest();
			}	catch	(NoSuchAlgorithmException	|	IOException	e)	{
						//	Catches	both:	algorithm	and	file	I/O	exceptions
						e.printStackTrace();
			}
			return	bToH(hashedBytes);	 	 //	Convert	bytes	to	hex	string
}	//	hashAFile()

    public	static	void main(String[]	args)	{
			String	msg	="It was the best of times, it was the worst of times.";
			String	hashValue	= hashString(msg,"MD5");
			System.out.println(hashValue+" "+msg);
            String hashValue2 = hashString(msg,"SHA-256");
			System.out.println(hashValue2+" "+msg);
            String name1 = "TwoCities.txt";
            String name2 = "Anassa.txt";
            String name3 = "moby10b.txt";
            String name4 = "SpongeBob.png";
            var reader = new md5();
            /*String fileValue = reader.hashAFile(name1,"MD5");
            System.out.println(fileValue+" "+name1);
            String fileValue2 = reader.hashAFile(name2,"MD5");
            System.out.println(fileValue2+" "+name2);
            String fileValue3 = reader.hashAFile(name3,"MD5");
            System.out.println(fileValue3+" "+name3);
            String fileValue4 = reader.hashAFile(name4,"MD5");
            System.out.println(fileValue4+" "+name4);
            
            
            String fileValue = reader.hashAFile(name1,"SHA-256");
            System.out.println(fileValue+" "+name1);
            String fileValue2 = reader.hashAFile(name2,"SHA-256");
            System.out.println(fileValue2+" "+name2);
            String fileValue3 = reader.hashAFile(name3,"SHA-256");
            System.out.println(fileValue3+" "+name3);
            String fileValue4 = reader.hashAFile(name4,"SHA-256");
            System.out.println(fileValue4+" "+name4);
            */
            String file1 = "/home/gtowell/Public/337/Lab10/shattered-1.pdf";
            String file2 = "/home/gtowell/Public/337/Lab10/shattered-2.pdf";
            System.out.println();
            String fileValue71 = reader.hashAFile(file1,"MD5");
            System.out.println("MD5:"+fileValue71+" "+file1);
            String fileValue72 = reader.hashAFile(file1,"SHA-256");
            System.out.println("SHA-256:"+fileValue72+" "+file1);
            String fileValue73 = reader.hashAFile(file1,"SHA1");
            System.out.println("SHA1:"+fileValue73+" "+file1);
            String fileValue74 = reader.hashAFile(file2,"MD5");
            System.out.println("MD5:"+fileValue74+" "+file2);
            String fileValue75 = reader.hashAFile(file2,"SHA-256");
            System.out.println("SHA-256:"+fileValue75+" "+file2);
            String fileValue76 = reader.hashAFile(file2,"SHA1");
            System.out.println("SHA1:"+fileValue76+" "+file2);


    }	//	main()
}